### PR TITLE
[YUNIKORN-1575] [shim] Ability to specify guaranteedResources via namespace annotations

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
### What is this PR for?
The shim side change for the ability to specify guaranteedResources via namespace annotations.

The interface change has been merged. https://github.com/apache/yunikorn-scheduler-interface/pull/94

The core side changes will following!

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/YUNIKORN-1575
* Put link here, and add [YUNIKORN-*Jira number*] in PR title, eg. `[YUNIKORN-2] Gang scheduling interface parameters`

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/cache/context.go [hunk 1]
```diff
@@ -706,6 +706,15 @@ func (ctx *Context) updateApplicationTags(request *interfaces.AddApplicationRequ
 			request.Metadata.Tags[siCommon.AppTagNamespaceResourceQuota] = string(quotaStr)
 		}
 	}
+
+	// add guaranteed resource info as an app tag
+	guaranteedResource := utils.GetNamespaceGuaranteedFromAnnotation(namespaceObj)
+	if guaranteedResource != nil && !common.IsZero(guaranteedResource) {
+		if guaranteedStr, err := json.Marshal(guaranteedResource); err == nil {
```
Comment (pbacsko) [inline pkg/cache/context.go:713]: Nit: `guaranteedStr` is not a string, it's a byte array. Just call it `guaranteed`.

File: pkg/common/utils/utils.go [hunk 2]
```diff
@@ -142,6 +142,21 @@ func PodUnderCondition(pod *v1.Pod, condition *v1.PodCondition) bool {
...
+// get namespace guaranteed resource from namespace annotation
+func GetNamespaceGuaranteedFromAnnotation(namespaceObj *v1.Namespace) *si.Resource {
+	// retrieve guaranteed resource info from annotations
+	namespaceGuaranteed := GetNameSpaceAnnotationValue(namespaceObj, constants.NamespaceGuaranteed)
+	var namespaceGuaranteedMap map[string]string
+	err := json.Unmarshal([]byte(namespaceGuaranteed), &namespaceGuaranteedMap)
+	if err != nil {
+		log.Logger().Warn("Unable to process namespace.guaranteed annotation",
+			zap.String("namespace", namespaceObj.Name),
+			zap.String("namespace.guaranteed val is", namespaceGuaranteed))
```
Comment (pbacsko) [inline pkg/common/utils/utils.go:154]: `namespace.guaranteed val is` -> just simply `namespace.guaranteed`

**Review by pbacsko — CHANGES_REQUESTED** [review]
Two minor changes

File: pkg/common/utils/utils.go (hunk 2 above)
Comment (zhuqi-lucas) [reply pkg/common/utils/utils.go]: Addressed in latest PR, thanks @pbacsko .

**Comment by zhuqi-lucas** [comment]
> Two minor changes

Thanks for review @pbacsko , addressed comments.

**Comment by zhuqi-lucas** [comment]
Error: ../../../go/pkg/mod/github.com/apache/yunikorn-core@v0.0.0-20230614080512-68a82820f0d7/pkg/events/events.go:31:3: unknown field GroupID in struct literal of type si.EventRecord
Error: ../../../go/pkg/mod/github.com/apache/yunikorn-core@v0.0.0-20230614080512-68a82820f0d7/pkg/events/events.go:32:3: unknown field Reason in struct literal of type si.EventRecord


Need to update core side to update SI interface too, i will create a jira for this.

**Comment by zhuqi-lucas** [comment]
The e2e test fail is not related, it's the timeout error.

File: pkg/cache/context.go [hunk 3]
```diff
@@ -702,10 +702,19 @@ func (ctx *Context) updateApplicationTags(request *interfaces.AddApplicationRequ
...
 		}
 	}
+
+	// add guaranteed resource info as an app tag
+	guaranteedResource := utils.GetNamespaceGuaranteedFromAnnotation(namespaceObj)
+	if guaranteedResource != nil && !common.IsZero(guaranteedResource) {
+		if guaranteed, err := json.Marshal(guaranteedResource); err == nil {
+			request.Metadata.Tags[siCommon.AppTagNamespaceResourceGuaranteed] = string(guaranteed)
+		}
+	}
```
Comment (pbacsko) [inline pkg/cache/context.go:716]: Please add test coverage to `TestAddApplicationsWithTags()`, there is a dummy lister and existing namespace object which you can re-use.

**Review by pbacsko — CHANGES_REQUESTED** [review]
Please enhance coverage (lines 713-715).

**Comment by zhuqi-lucas** [comment]
@pbacsko 
One more question, how can we add the unit test to cover the log, for example, the log here, how we can improve the unit test to cover this, using log hook? Thanks!
```
	if err != nil {
		log.Logger().Warn("Unable to process namespace.guaranteed annotation",
			zap.String("namespace", namespaceObj.Name),
			zap.String("namespace.guaranteed is", namespaceGuaranteed))
		return nil
	}
```

**Comment by pbacsko** [comment]
> @pbacsko One more question, how can we add the unit test to cover the log, for example, the log here, how we can improve the unit test to cover this, using log hook? Thanks!
> 
> ```
> 	if err != nil {
> 		log.Logger().Warn("Unable to process namespace.guaranteed annotation",
> 			zap.String("namespace", namespaceObj.Name),
> 			zap.String("namespace.guaranteed is", namespaceGuaranteed))
> 		return nil
> 	}
> ```

I think it's an overkill to capture log output. If you can check some side effect (return value, mock state) then do it, otherwise it's fine to skip. There are paths in the code where it's difficult to simulate an error condition, in those cases, we don't have coverage.

**Review by pbacsko — APPROVED** [review]
+1 LGTM.
You can add more tests if you think it's necessary.

**Comment by zhuqi-lucas** [comment]
It makes sense, thanks @pbacsko for explain!

**Pull request closed without merging**